src/ema/math.rs

```diff
@@ -0,0 +1,114 @@
+use crate::types::{Balance, Price};
+
+use sp_arithmetic::{
+    traits::{One, Saturating},
+    FixedPointNumber, FixedU128,
+};
+
+/// Calculate the iterated exponential moving average for the given prices.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_price_ema(iterations: u32, prev: Price, incoming: Price, smoothing: FixedU128) -> Price {
+    let (exp_smoothing, exp_complement) = exp_smoothing_and_complement(smoothing, iterations);
+    price_ema(prev, exp_complement, incoming, exp_smoothing)
+}
+
+/// Calculate the iterated exponential moving average for the given balances.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_balance_ema(iterations: u32, prev: Balance, incoming: Balance, smoothing: FixedU128) -> Balance {
+    let (exp_smoothing, exp_complement) = exp_smoothing_and_complement(smoothing, iterations);
+    balance_ema(prev, exp_complement, incoming, exp_smoothing)
+}
+
+/// Calculate the iterated exponential moving average for the given volumes.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_volume_ema(
+    iterations: u32,
+    prev: (Balance, Balance, Balance, Balance),
+    incoming: (Balance, Balance, Balance, Balance),
+    smoothing: FixedU128,
+) -> (Balance, Balance, Balance, Balance) {
+    let (exp_smoothing, exp_complement) = exp_smoothing_and_complement(smoothing, iterations);
+    volume_ema(prev, exp_complement, incoming, exp_smoothing)
+}
+
+/// Calculate the smoothing factor and its complement for a given combination of period and
+/// iterations.
+pub fn exp_smoothing_and_complement(smoothing: FixedU128, iterations: u32) -> (FixedU128, FixedU128) {
+    debug_assert!(smoothing <= Price::one());
+    let complement = Price::one() - smoothing;
+    // in order to determine the iterated smoothing factor we exponentiate the complement
+    let exp_complement = complement.saturating_pow(iterations as usize);
```

Comment:
Does saturating_pow round everything down? Do we know anything about its accuracy guarantees?

Reply:
[`saturating_pow` uses `checked_mul`](https://github.com/paritytech/substrate/blob/ce10b9f29353e89fc3e59d447041bb29622def3f/primitives/arithmetic/src/fixed_point.rs#L752-L769) under the hood [which rounds down](https://github.com/paritytech/substrate/blob/ce10b9f29353e89fc3e59d447041bb29622def3f/primitives/arithmetic/src/fixed_point.rs#L860-L865):

```Rust
multiply_by_rational_with_rounding(
					lhs.value,
					rhs.value,
					Self::DIV as u128,
					Rounding::from_signed(SignedRounding::Minor, negative),
				)
```

Reply:
This seems pretty likely to introduce a long-term bias to the oracles which may be a problem. We'll probably want to at least test this, by testing the effect of updating the oracle every period for a bunch of iterations against one update spanning a bunch of periods... This would be where the bias would show up I think.